libs/header/sync/ranges.go

```diff
@@ -83,6 +98,26 @@ func (rs *ranges[H]) FirstRangeWithin(start, end uint64) (*headerRange[H], bool)
 	return nil, false
 }
 
+// FindAllRangesWithin returns all headerRanges that apply the requested range.
+// Returned ranges contain `start` within their bounds.
+// NOTE: ranges do not need to contain `end`.
+// There could be multiple headerRanges:
+// start = 2 , end 40
+// headerRange0 = [3;20];
+// headerRange1 = [30; 35];
+func (rs *ranges[H]) FindAllRangesWithin(start, end uint64) []*headerRange[H] {
```

Comment:
* We allocate here an additional slice, unnecessary
* We leak empty ranges here. The canonical First method below was responsible for cleaning them

Reply:
no more leaks. No functions - no problems 😄